sonic_data_client/mixed_db_client.go

```diff
@@ -220,15 +243,132 @@ func (c *MixedDbClient) DbDelTable(table string, key string) error {
 	return nil
 }
 
+// For example, the GNMI path below points to DASH_QOS table in the DPU_APPL_DB database for dpu0:
+// /DPU_APPL_DB/dpu0/DASH_QOS
+// The first element of the GNMI path is the target database name, which is DPU_APPL_DB in this case.
+// The second element is the container name, which is dpu0.
+// The third element is the node name, which is DASH_QOS.
+// ParseDatabase is used get database target and SonicDBKey from GNMI prefix and path
+func (c *MixedDbClient) ParseDatabase(prefix *gnmipb.Path, paths []*gnmipb.Path) (string, swsscommon.SonicDBKey, error) {
+	if len(paths) == 0 {
+		return "", nil, status.Error(codes.Unimplemented, "No valid path")
+	}
+	target := ""
+	namespace := sdcfg.SONIC_DEFAULT_NAMESPACE
+	container := sdcfg.SONIC_DEFAULT_CONTAINER
+	for _, path := range paths {
+		if path.GetElem() != nil {
+			elems := path.GetElem()
+			if prefix != nil {
+				elems = append(prefix.GetElem(), elems...)
+			}
+			if elems == nil {
+				return "", nil, status.Error(codes.Unimplemented, "No valid elem")
+			}
+			if c.namespace_cnt > 1 || c.container_cnt > 1 {
+				if len(elems) < 2 {
+					return "", nil, status.Error(codes.Unimplemented, "Invalid elem length")
+				}
+			}
+			// Get target from the first element
+			if target == "" {
+				target = elems[0].GetName()
+			} else if target != elems[0].GetName() {
+				return "", nil, status.Error(codes.Unimplemented, "Target conflict in path")
+			}
+			if c.namespace_cnt > 1 {
+				// Get namespace from the second element
```

Comment:
>namespace [](http://example.com/codeflow?start=11&length=9)

Are you assuming namespace and container could not used together? This assumption may be not necessary.

Reply:
@zegan Could you check "sonic-swss-common does not support namespace and container together"? If yes, seems a bug.

Reply:
I feel the API design of sonic-swss-common should support namespace and container together.
But right now, we don't have this scenario with multi DPU + multi ASIC. So, I feel the implementation may have some issues. Not only swss-common but also other components cannot support this mixed mode.

So, IMHO, the interface design should support this mixed mode, but the implementation isn't necessary.

Reply:
I have updated, now we can support SmartSwitch with a multi-asic NPU.

Reply:
closed.